onnx/__init__.py

```diff
@@ -14,59 +14,110 @@
-    '''
+def _load_bytes(obj):
     if hasattr(obj, 'read') and callable(obj.read):
         s = obj.read()
     else:
         with open(obj, 'rb') as f:
             s = f.read()
-    return load_from_string(s)
+    return s
+
+
+def _save_bytes(s, f):
+    if hasattr(f, 'write') and callable(f.write):
+        f.write(s)
+    else:
+        with open(f, 'wb') as writable:
+            writable.write(s)
 
 
-def load_from_string(s):
+def serialize(proto):
     '''
-    Loads a binary string that stores onnx model
+    Serialize a in-memory proto to bytes
 
     @params
-    Takes a string object containing protobuf
-    @return ONNX ModelProto object
+    proto is a in-memory proto, such as a ModelProto, TensorProto, etc
+
+    @return
+    Serialized proto in bytes
+    '''
+    if isinstance(proto, bytes):
+        return proto
+    elif hasattr(proto, 'SerializeToString') and callable(proto.SerializeToString):
+        return proto.SerializeToString()
+    else:
+        raise ValueError('No SerializeToString method is detected. '
+                         'neither obj is bytes.\ntype is {}'.format(type(proto)))
+
+
+def deserialize(s, obj):
+    '''
+    Parse bytes into a in-memory proto
+
+    @params
+    s is bytes containing serialized proto
+    obj can be a class object, such as ModelProto, also can be a proto instance to be filled in
+
+    @return
+    The proto instance filled in by s
     '''
-    model = ModelProto()
-    decoded = model.ParseFromString(s)
-    # in python implementation ParseFromString returns None
+    if not isinstance(s, bytes):
+        raise ValueError('Parameter s must be bytes, but got type: {}'.format(type(s)))
+
+    if callable(obj):
+        obj = obj()
+    if not (hasattr(obj, 'ParseFromString') and callable(obj.ParseFromString)):
+        raise ValueError('No ParseFromString method is detected. '
+                         '\ntype is {}'.format(type(obj)))
+
+    decoded = obj.ParseFromString(s)
     if decoded is not None and decoded != len(s):
         raise google.protobuf.message.DecodeError(
             "Protobuf decoding consumed too few bytes: {} out of {}".format(
                 decoded, len(s)))
-    return model
+    return obj
 
 
-def _save(model, writable):
-    if isinstance(model, ModelProto):
-        writable.write(model.SerializeToString())
-    elif isinstance(model, str):
-        writable.write(model)
-    else:
-        raise ValueError('Model is neither ModelProto nor str.\n{}'.format(model))
+def load(obj, cls=None):
+    '''
+    Loads a binary protobuf into memory
+
+    @params
+    obj can be a file-like object (has "read" function) or a string containing a file name
+    cls is a class object, such as ModelProto, TensorProto, etc.
+    If cls is not specified, loads obj as ModelProto
+
+    @return
+    Loaded in-memory proto object
+    '''
+    s = _load_bytes(obj)
+    return load_from_string(s, cls)
+
 
+def load_from_string(s, cls=None):
```

Comment:
why not default to ModelProto?

Reply:
I feel bad to use any mutable object as default values